Use exit status instead of IsGitErrorMessage()

gerhardol opened the pull request:
Preparation for #9056 

## Proposed changes

ExecutableExtensions: CommandCache for Execute()
Previously caching was possible only for GetOutput()

Instead of parsing the combined standard/error output for 'fatal' or 'error:' in IsGitErrorMessage(), use the executable exit status.
The standard usage is to add a a fake GitStatusItem to show the failed command for the user in FileStatusList.
This handling is done _after_ standard output is parsed, as it is currently. 

Note that changes like GitConfigurationException makes it hard to reach these codepaths (you get constant popups, cannot do anything) so testing this is hard.

## Test methodology <!-- How did you ensure quality? -->

Manual, some commands are tested though

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitCommands/Git/GitModule.cs:
```diff
@@ -2510,14 +2510,12 @@ public IReadOnlyList<GitItemStatus> GetTreeFiles(ObjectId commitId, bool full)
-
-            List<GitItemStatus> result = _getAllChangedFilesOutputParser.Parse(output).ToList();
-            if (IsGitErrorMessage(output))
+            var exec = _gitExecutable.Execute(GitCommandHelpers.GetAllChangedFilesCmd(excludeIgnoredFiles, untrackedFiles));
```
Please use type instead of var

gerhardol replied:
done

RussKie reviewed (commented):
LGTM
How did you test these?

gerhardol commented:
> LGTM
> How did you test these?

Some commands have tests (but not really related to errors).

Manual usage, adding errors (bad config like when IsGitError was added), for some commands. (GitConfigurationException makes this much harder to use, you do not get much functionality.)
The handling is unchanged, some processing also if there are errors - this could be changed but are harder to review.
It may be that we should remove some of the handling with #9056 too, then GE is unusable if there is any unexpected Git error.

RussKie commented:
👍

On Sun, 11 Apr. 2021, 9:53 am Gerhard Olsson, ***@***.***>
wrote:

> LGTM
> How did you test these?
>
> Some commands have tests (but not really related to errors).
>
> Manual usage, adding errors (bad config like when IsGitError was added),
> for some commands. (GitConfigurationException makes this much harder to
> use, you do not get much functionality.)
> The handling is unchanged, some processing also if there are errors - this
> could be changed but are harder to review.
> It may be that we should remove some of the handling with #9056
> <https://github.com/gitextensions/gitextensions/pull/9056> too, then GE
> is unusable if there is any unexpected Git error.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9059#issuecomment-817218537>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXSH3QNTAGAGKS52NMLTIDQIVANCNFSM42WUMBZQ>
> .
>

gerhardol commented:
Updated with: ExecutableExtensions: CommandCache for Execute()

gerhardol commented:
@mstv This affects #9056, I suggest this is merged first

mstv commented on GitCommands/Git/ExecutableExtensions.cs:
```diff
@@ -363,9 +374,13 @@ string ComposeOutput()
 
+            if (cache is not null && await exitTask == 0)
+            {
+                cache.Add(arguments, outputBuffer.ToArray(), errorBuffer.ToArray());
```
We must be careful not to cache too much.
I guess that I have seen diffs with Index and WD not updated correctly until I reloaded everything in the past month.

gerhardol replied:
Sidenote to this PR:
If something is cached, a refresh will not update automatically, the cache contents is kept.
So what you saw is likely that the file lists or the contents of a diff is not updated automatically at changes.
I would like to have notification about changes in the diff tab and FormCommit similar to the Browse refresh (FileStatusMonitor) to detect this.

However, I want a cache for git-status too, but that is a new functionality, the cache will need quite heavy invalidation policy...

mstv commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3032,21 +3022,20 @@ public IReadOnlyList<string> GetAllBranchesWhichContainGivenCommit(ObjectId obje
-            if (IsGitErrorMessage(output))
+                };
+            ExecutionResult exec = _gitExecutable.Execute(args);
+            if (!exec.ExitedSuccessfully)
             {
                 return Array.Empty<string>();
             }
```
What do we do with the error message in this case?
After #9056, we should throw an `ExternalOperationException` here, which must be caught by the caller and be displayed in the UI element.

gerhardol replied:
adding comment

mstv commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3074,14 +3063,13 @@ public IReadOnlyList<string> GetAllBranchesWhichContainGivenCommit(ObjectId obje
-
-            if (IsGitErrorMessage(output))
+            ExecutionResult exec = _gitExecutable.Execute($"tag --contains {objectId}");
+            if (!exec.ExitedSuccessfully)
             {
                 return Array.Empty<string>();
             }
```
ditto

gerhardol replied:
adding comment

mstv commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3115,11 +3103,12 @@ public IReadOnlyList<string> GetAllTagsWhichContainGivenCommit(ObjectId objectId
              */
 
-            if (IsGitErrorMessage(output))
+            if (!exec.ExitedSuccessfully)
             {
                 return null;
             }
```
ditto

gerhardol replied:
adding comment

mstv commented on GitCommands/Git/GitModule.cs:
```diff
@@ -4146,24 +4135,11 @@ public bool StopTrackingFile(string filename)
-        public static bool IsGitErrorMessage(string gitOutput)
-        {
-            return Regex.IsMatch(gitOutput, @"^\s*(error:|fatal)");
+            ExecutionResult exec = _gitExecutable.Execute(args);
+            return !exec.ExitedSuccessfully
+                ? exec.StandardOutput.TrimEnd()
+                : null;
```
ditto

gerhardol replied:
Exit > 0 is normal if no tag exists where the command was started, must not raise errors

mstv reviewed (commented):
Looks good, Just a few nits.
How ´do you trigger the git errors?

gerhardol commented:
> How ´do you trigger the git errors?

You cannot easily trigger most of them.
See comment https://github.com/gitextensions/gitextensions/pull/9059#issuecomment-817218537

This PR is not really trying to improve the handling, it is a preparation for #9056 where commands currently handling Git errors but the handling was not easily visible and was not handled in the draft #9056.
 
The review comments about error messages not displayed are like that, the messages were not displayed before, at some point it was not fatal, compare to git-rev-parse where an error is normal. Maybe these situations should be reconsidered. That was not my goal for this PR though.

gerhardol commented:
Rebased (on #9058) and added a fixup with an explanation where the error message should be presented (removing it would have been a regression).
See below, the "fatal:" info is useful information.
(This is something I see quite often when working with submodules and worktrees).

```
Submodule Externals/EasyHook Change

From:	5183aa53a286771f10b5828053e9474a30dc4db1

To:	0296f95158a7da669b44aaa264d8793c478ef648
					1 year ago (2020-01-16 11:56:13)
		Remove targets, replaced by super's targets/props

Type: Unknown

Differences:

fatal: bad object 5183aa53a286771f10b5828053e9474a30dc4db1
```

I need https://github.com/gitextensions/gitextensions/issues/8914#issuecomment-790080907
to run this also on my home PC

gerhardol commented on GitCommands/Git/ExecutableExtensions.cs:
```diff
@@ -340,10 +342,19 @@ string ComposeOutput()
         {
             outputEncoding ??= _defaultOutputEncoding.Value;
 
+            if (cache is not null && cache.TryGet(arguments, out var cacheOutput, out var cacheError))
```
ok

gerhardol commented on GitCommands/Git/ExecutableExtensions.cs:
```diff
@@ -340,10 +342,19 @@ string ComposeOutput()
 
+            if (cache is not null && cache.TryGet(arguments, out var cacheOutput, out var cacheError))
+            {
+                return new ExecutionResult(
+                    CleanString(stripAnsiEscapeCodes, EncodingHelper.DecodeString(cacheOutput, null, ref outputEncoding)),
+                    CleanString(stripAnsiEscapeCodes, EncodingHelper.DecodeString(null, cacheError, ref outputEncoding)),
+                    0);
```
ok

gerhardol commented on GitCommands/Git/ExecutableExtensions.cs:
```diff
@@ -363,9 +374,13 @@ string ComposeOutput()
-
             var exitCode = await process.WaitForExitAsync();
 
+            if (cache is not null && await exitTask == 0)
```
ok

mstv reviewed (approved):
Thank you.

ivangrek commented on GitCommands/Git/GitModule.cs:
```diff
@@ -2725,10 +2725,20 @@ public bool IsDirtyDir()
+            if (!file.IsTracked || (file.Staged == StagedStatus.Index && file.IsNew))
+            {
+                // File is not in Git
+                return null;
```
return Task.FromResult(null)

gerhardol replied:
The task is not null, it returns a null string. What is the point of writing that with more text?

ivangrek replied:
If task is not null it is ok.
I'll take a closer look. 😄

ivangrek replied:
I don't know why we are using `ConfigureAwaitRunInline()` there. But we could try to get the content in parallel for two files.

gerhardol replied:
This is in #9058, I based this PR on #9058 to prepare the merge of this, so discussed there.

RussKie replied:
My first reaction was too to suggest `Task.FromResult(null)`. But since there's an `await` below, this is not necessary, and distorts the intent.

ivangrek commented on GitUI/AutoCompletion/CommitAutoCompleteProvider.cs:
```diff
@@ -149,19 +149,21 @@ private static IEnumerable<string> ReadOrInitializeAutoCompleteRegexes()
-            {
-                return content;
+                var content = await module.GetFileContentsAsync(file).ConfigureAwaitRunInline();
+                if (content is not null)
```
Why?

mstv replied:
This is an early exit. Else it is tried another way below. So why "Why?"?

ivangrek replied:
Ok.

ivangrek commented on GitCommands/Git/ExecutableExtensions.cs:
```diff
@@ -376,7 +376,7 @@ string ComposeOutput()
             var error = outputEncoding.GetString(errorBuffer.GetBuffer(), 0, (int)errorBuffer.Length);
             var exitCode = await process.WaitForExitAsync();
```
I think we need to use `await exitTask` there. 
To prevent second run.

mstv replied:
Done in line 373: `await Task.WhenAll(outputTask, errorTask, exitTask);`.

ivangrek replied:
![image](https://user-images.githubusercontent.com/3169265/114910264-57910600-9e26-11eb-8379-7ab6be721507.png)

ivangrek replied:
Also.
https://www.meziantou.net/process-waitforexitasync-doesn-t-behave-like-process-waitforexit.htm

mstv replied:
> I think we need to use `await exitTask` there.
To prevent second run.

Thank you, I got it now. (I think in this case, it will not do harm and will not execute very much twice. Though beware of early signs, i.e. it should be changed.)

The .NET 5 issue does not affect this code because the output tasks are awaited explicitly.

gerhardol replied:
This was not changed in this PR, hesitant to change...

RussKie commented:
@gerhardol merge at will

gerhardol commented:
Rebased and squashed with an indentation change in GitModule.
Merging despite two partly open discussions, not really this PR:
* Task null return that is part of #9058 (mstv tried to stop the bot merge so I did not try that myself, but it failed...). RussKie resolved the conversation.
* Execute() await handling. Existing code

gerhardol commented:
@msftbot merge in 12 hours

mstv commented:
> mstv tried to stop the bot merge

I stopped it because of the new review comment - by removing the `auto-merge` label, which worked.
After you had rebuted it, I re-added the label and #9058 was merged as planned.
